### ISSUE TITLE
Anchor in URL does not change when changing the combination ID in the url

### ISSUE THREAD

**Issue by colinegin** [PrestaShop/PrestaShop#18546]
### Describe the bug
When you change the id of combination in the url, you will be redirect to the default combination but the url won't be corrected/changed 

### Expected behavior
these parameters in the url should be replaced by those of the default combination

### Steps to reproduce

1. if we go to a product with combination
2. the default combination is displayed
3. If we change combination the URL is modified
4. if we change the combination ID indeed we are well redirected and the default combination is displayed
5. but the previously modified parameters are still in the URL see screen record.
6. Shouldn't these parameters be replaced by those of the default combination?

**Screen Record**
https://drive.google.com/file/d/1vnvafdkeOHsp7FN4Nz5Tmf0wMJH3BwXL/view?usp=sharing

### Additional infos

- PS version: 8.1.x & 9.0.x
- PHP version: 8.1

**Comment by jolelievre** [comment]
See this comment for more detail: https://github.com/PrestaShop/PrestaShop/pull/18493#issuecomment-611504504

tl;dr;
- keeping the anchor is a logical browser side feature
- we need to decide if the anchor affects the combination displayed in the page
- if it does affect it, does it have influence on SEO? (javascript front update or PHP server side rendering)

**Comment by kpodemski** [comment]
Everything after the anchor is irrelevant for both things. To be honest i'm not sure why anchor is still used if we have combination id in URL 🤔

**Comment by colinegin** [comment]
I guess it can help the user know what elements of the combination (ex : color black and size S) will be shown directly from the URL

**Comment by boo-code** [comment]
Still the case on 9.2.x, and the reason is narrower than the discussion assumed, so it is worth pinning down.

Measured. A product URL carrying a combination looks like:

```
http://…/1-1-hummingbird-printed-t-shirt.html#/size-s/color-white
```

Requesting one with an invalid combination id, `1-99999-hummingbird-printed-t-shirt.html`, the controller does redirect, to:

```
http://…/1-hummingbird-printed-t-shirt.html
```

That target has **no fragment**. `ProductController::canonicalRedirection()` clears `$this->id_product_attribute` when the combination is invalid and then builds the target with `getProductLink(..., $this->id_product_attribute)`, and `Link::getProductLink()` only appends an anchor when an attribute id is given - `$anchor = $addAnchor && $idProductAttribute ? $product->getAnchor(...) : ''`. So the redirect location is fragment-less by construction.

That is what keeps the stale anchor on screen. RFC 7231 section 7.1.2 says that when the redirect target has no fragment, the user agent applies the original request's fragment to it. The browser is not ignoring PrestaShop, it is doing what it is told when told nothing.

Which also means this is fixable server side, contrary to what "keeping the anchor is a logical browser side feature" might suggest: a redirect target that carries its own fragment overrides the original one. So the choice is not between PHP and JavaScript, it is about what the fragment should say:

- redirecting to the default combination's own URL, `1-1-…html#/size-s/color-white`, is self consistent and is what the report asks for, but it changes the redirect target from the plain product URL to a combination URL, which is squarely @jolelievre's SEO question;
- emitting an empty fragment on the existing target only clears the stale anchor without asserting anything about combinations, which needs no decision about what the anchor means, but a bare `#` in a `Location` header is not something the codebase does anywhere today.

Both leave @kpodemski's larger question untouched, namely whether the anchor is still earning its place now that the combination id is in the path. I have not built either; the measurement above is the part that seemed to be missing.
